File: src/main/java/org/tindalos/principle/domain/expectations/Layering.java
package org.tindalos.principle.domain.expectations;

import java.util.List;

/**
 * Layering architecture expectation.
 *
 * <p>Layering enforces that dependencies between architectural layers should only flow in one
 * direction - from higher layers to lower layers. This prevents circular dependencies between
 * layers and maintains a clear architectural structure.</p>
 *
 * <p>In a typical layered architecture:</p>
 * <ul>
 *   <li><b>Presentation Layer</b> → depends on Business Layer</li>
 *   <li><b>Business Layer</b> → depends on Data Layer</li>
 *   <li><b>Data Layer</b> → has no dependencies (lowest layer)</li>
 * </ul>
 *
 * <p>Violations occur when:</p>
 * <ul>
 *   <li>A lower layer depends on a higher layer (upward dependency)</li>
 *   <li>Layers are bypassed (e.g., Presentation directly depends on Data, skipping Business)</li>
 * </ul>
 *
 * <p>The layers list defines the order from highest to lowest. For example:
 * {@code ["infrastructure", "app", "domain"]} where infrastructure is the highest layer
 * and domain is the lowest.</p>
 *
 * @param layers the list of layer names in order from highest to lowest
 * @param violationThreshold the maximum number of layering violations allowed (default is 0)
 */
public record Layering(List<String> layers, int violationThreshold) implements IntThresholder {

    public Layering() {
        this(List.of(), 0);
    }
    public Layering {
        layers = List.copyOf(layers);
    }
}

File: src/main/scala/org/tindalos/principle/domain/expectations/Layering.scala
package org.tindalos.principle.domain.expectations

import org.tindalos.principle.domain.expectations.exception.InvalidConfigurationException
import org.tindalos.principle.domain.util.ListConverter
import org.tindalos.principle.domain.util.ListConverter

case class Layering(var layers: List[String] = List.empty, threshold: Int = 0) extends Thresholder {

  def setLayers(theLayers: java.util.List[String]) = {
    if (theLayers == null || theLayers.isEmpty)
      throw new InvalidConfigurationException("Invalid layering expectation! No layers specified!")
    layers = ListConverter.convert(theLayers)
  }

}